src/material/core/ripple/ripple-renderer.ts

```diff
@@ -176,6 +176,10 @@ export class RippleRenderer implements EventListenerObject {
     if (!animationForciblyDisabledThroughCss && (enterDuration || animationConfig.exitDuration)) {
       this._ngZone.runOutsideAngular(() => {
         ripple.addEventListener('transitionend', () => this._finishRippleTransition(rippleRef));
+        // If the transition is cancelled (e.g. due to DOM removal), we destroy the ripple
+        // directly as otherwise we would keep it part of the ripple container forever.
+        // https://www.w3.org/TR/css-transitions-1/#:~:text=no%20longer%20in%20the%20document.
+        ripple.addEventListener('transitioncancel', () => this._destroyRipple(rippleRef));
```

Comment:
Is this going to fire if the transition never starts?

Reply:
If there is no transition, then we will not even hit that code path (see conditions above and check belkow).

In other cases, if the transition is set by the renderer, the outcome will either be `transitionend` or `transitioncancel`. do you have something else in mind? There is an edge-case where the ripple container is already out of the DOM, but that seems unexpected anyway..

Reply:
I see. I guess I'm concerned, because `transitioncancel` is marked as experimental on MDN. Since we already know the transition duration, couldn't we have a `setTimeout` as a backup if the `transitionend` doesn't fire on time + some amount of buffer? That's how we handle such cases in the `cdk/drag-drop`.

Reply:
I thought about doing that, also remembering some code in the CDK overlay, but it felt more ideal avoiding two additional timers for just this edge case.

Yes, I saw this being marked as experimental but thought giving it a shot given representatives of all major browsers having participated in the CSS Transition spec, its latest revision being unchanged since 2018, and browser support being petty good. I'm not 100% sure what state this spec is in though. i.e why.

That's also why I asked you to check for usages in g3.

Reply:
I was imagining that we'd clear the timeouts in `transitionend` so they shouldn't be a problem in tests.

Reply:
yeah, not all tests would flush both fade-in, fade-out though, and would just remove the test fixture from DOM. I was not necessarily only thinking about tests, but rather about the fact that we'd schedule two macro-tasks (no necessarily expensive though) and that there would be a small delay for the "destroy" procedure in case of a DOM removal.

In any case though, I do not feel super strongly at all. Happy to go with either way. Maybe @andrewseguin has some thoughts on this.

Reply:
I understand the concern about relying on experimental behavior, but it does seem stable and it's hard to imagine browsers breaking this. In the worst case, an issue pops up saying that its broken on a browser and at that point creating the fix should be straightforward. I think we should move forward with this simpler patch